### PR TITLE
Make node timers more efficient

### PR DESCRIPTION
**Pull request by Ekdohibs**
This pull request makes node timers more efficient by using a map from times to trigger node timers to the timers to trigger, and thus looking only at the timers that need to each step.

### CODE REVIEW

**Comment by sofar** [comment]
I spotted a few ```if(``` cases and a few ```{}``` pairs that can be removed (if() followed by a single statement.

Other than that no comments, I'll try and see if I can test this with some significant amount of nodetimers from my other PRs to minetest_game.

**Comment by paramat** [comment]
Added to milestone as we're considering making the change to nodetimer crops in MTGame.

**Comment by paramat** [comment]
Is there time to get this stable for release, or could it introduce subtle bugs?

**Comment by sofar** [comment]
I'll try and test this one tonight with various nodetimer patches I have made and pending.

**Comment by sofar** [comment]
@paramat Either this works, or it totally breaks.

I'm testing this right now, and I see no issues with it. Tested NodeTimer crops, furnaces, pressure plates. Everything seems to be in order.

**Comment by paramat** [comment]
Removed milestone but this could probably go in release.

File: src/nodetimer.cpp [hunk 1]
```diff
@@ -103,52 +103,49 @@ void NodeTimerList::deSerialize(std::istream &is, u8 map_format_version)
...
-	for(std::map<v3s16, NodeTimer>::iterator
-			i = m_data.begin();
-			i != m_data.end(); ++i){
-		v3s16 p = i->first;
+	std::vector<NodeTimer> elapsed_timers;
+	m_time += dtime;
+	if (m_next_trigger_time == -1. || m_time < m_next_trigger_time) {
+		return elapsed_timers;
+	}
+	std::map<double, NodeTimer>::iterator i = m_timers.begin();
```
Comment (SmallJoker) [inline src/nodetimer.cpp]: MSVC 2010 compiling error, please change this to `std::multimap`

File: src/map.cpp [hunk 2]
```diff
@@ -2084,11 +2084,13 @@ NodeTimer Map::getNodeTimer(v3s16 p)
 		return NodeTimer();
 	}
 	NodeTimer t = block->m_node_timers.get(p_rel);
-	return t;
+	NodeTimer nt(t.timeout, t.elapsed, p);
+	return nt;
 }
 
-void Map::setNodeTimer(v3s16 p, NodeTimer t)
+void Map::setNodeTimer(NodeTimer t)
```
Comment (nerzhul) [inline src/map.cpp]: const NodeTimer &

File: src/mapblock.h [hunk 3]
```diff
@@ -488,9 +488,9 @@ class MapBlock /*: public NodeContainer*/
 		m_node_timers.remove(p);
 	}
 
-	inline void setNodeTimer(v3s16 p, NodeTimer t)
+	inline void setNodeTimer(NodeTimer t)
```
Comment (nerzhul) [inline src/mapblock.h]: const NodeTimer &

File: src/nodetimer.h [hunk 4]
```diff
@@ -36,15 +37,18 @@ class NodeTimer
 {
 public:
 	NodeTimer(): timeout(0.), elapsed(0.) {}
-	NodeTimer(f32 timeout_, f32 elapsed_):
-		timeout(timeout_), elapsed(elapsed_) {}
+	NodeTimer(v3s16 position_):
```
Comment (nerzhul) [inline src/nodetimer.h]: const v3s16 &

File: src/nodetimer.h [hunk 5]
```diff
@@ -54,37 +58,75 @@ class NodeTimer
...
 	~NodeTimerList() {}
 	
 	void serialize(std::ostream &os, u8 map_format_version) const;
 	void deSerialize(std::istream &is, u8 map_format_version);
 	
 	// Get timer
-	NodeTimer get(v3s16 p){
-		std::map<v3s16, NodeTimer>::iterator n = m_data.find(p);
-		if(n == m_data.end())
+	NodeTimer get(v3s16 p) {
```
Comment (nerzhul) [inline src/nodetimer.h]: const v3s16 &

File: src/nodetimer.h [hunk 6]
```diff
@@ -54,37 +58,75 @@ class NodeTimer
...
+				trigger_time, timer
+			));
+		m_iterators.insert(std::pair<v3s16, std::multimap<double, NodeTimer>::iterator>(p, it));
+		if (m_next_trigger_time == -1. || trigger_time < m_next_trigger_time)
+			m_next_trigger_time = trigger_time;
 	}
 	// Deletes old timer and sets a new one
-	void set(v3s16 p, NodeTimer t){
-		m_data[p] = t;
+	void set(NodeTimer timer) {
```
Comment (nerzhul) [inline src/nodetimer.h]: const NodeTimer &

File: src/map.h [hunk 7]
```diff
@@ -327,7 +327,7 @@ class Map /*: public NodeContainer*/
 	*/
 
 	NodeTimer getNodeTimer(v3s16 p);
-	void setNodeTimer(v3s16 p, NodeTimer t);
+	void setNodeTimer(NodeTimer t);
```
Comment (nerzhul) [inline src/map.h]: const NodeTimer &

**Comment by sofar** [comment]
I'm thinking we should just merge this and let it curate. :+1:

**Comment by paramat** [comment]
Can be merged now?

**Comment by Ekdohibs** [comment]
@paramat yes, it has been ready for some time now.

**Comment by paramat** [comment]
Ok will merge soon.

**Comment by paramat** [comment]
@Ekdohibs please could you fix the long lines and add missing spaces in 'if ()' and 'for ()', and also squash?

**Comment by Ekdohibs** [comment]
@paramat done.

**Comment by paramat** [comment]
559dd9946988cd35a7c26bcafe7d0f8c42dc547a

**Pull request closed without merging**